### ISSUE TITLE
Unable to run a non-system app

### ISSUE THREAD

**Issue by jaredly** [facebook/FBSimulatorControl#36]
I made a simple program to launch an app in a simulator, but it's not working :/ I started with the code example, in the Readme for launching safari, and that worked great.

Then I replaced the `configurationWithApplication` argument to be 

```objc
[FBSimulatorApplication
  applicationWithPath: @"/Path/to/My.app"
  error:&error]
```
Instead of the previous `[FBSimulatorApplication systemApplicationNamed:@"MobileSafari"]`

Inspecting the resulting, it seemed to have loaded the bundleId, binary path, etc. correctly, and it started the simulator, but it failed to install & launch my app.
Here's the error:
```
Domain=com.facebook.FBSimulatorControl Code=0 "Failed to launch application" UserInfo=0x10144bd00 {NSLocalizedDescription=Failed to launch application, launchd_is_running=true, launchd_subprocesses=(
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/sbin/syslogd | PID 34083",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/UserEventAgent (System) | PID 34084",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/FileProvider.framework/Support/fileproviderd | PID 34085",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/routined | PID 34086",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/MediaStream.framework/Support/mstreamd | PID 34087",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/MediaRemote.framework/Support/mediaremoted | PID 34088",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/IDSSimulatorSupport.framework/Support/ids_simd | PID 34089",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/installd | PID 34091",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/IDS.framework/identityservicesd.app/identityservicesd | PID 34092",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/TouchRemote.framework/Support/touchsetupd | PID 34093",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/CoreServices/AppleIDAuthAgent | PID 34094",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/Frameworks/WatchKit.framework/Support/companionappd | PID 34095",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/Frameworks/AssetsLibrary.framework/Support/assetsd | PID 34096",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/wcd | PID 34097",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/CoreServices/SpringBoard.app/SpringBoard | PID 34098",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/backboardd | PID 34099",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/locationd | PID 34101",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/assertiond | PID 34102",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/GenerationalStorage.framework/revisiond | PID 34103",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/CoreSimulator/RuntimeOverlay/usr/libexec/SimulatorBridge | PID 34104",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/configd_sim -d | PID 34105",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/CoreSimulator/RuntimeOverlay/usr/libexec/CoreSimulatorBridge | PID 34106",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/sbin/notifyd | PID 34107",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/sbin/cfprefsd daemon | PID 34108",
    "Process aslmanager | PID 34109",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/MobileGestaltHelper | PID 34110",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/nsurlsessiond | PID 34111",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/securityd | PID 34112",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/sbin/distnoted daemon | PID 34113",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/TelephonyUtilities.framework/callservicesd | PID 34114",
    "Process /usr/libexec/pkd -d/tmp/PlugInKit-Annotations | PID 34116",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/TCC.framework/tccd | PID 34118",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/ApplePushService.framework/apsd | PID 34119",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/networkd | PID 34120",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/MobileContainerManager.framework/Support/containermanagerd | PID 34122",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/mobileassetd | PID 34123",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/CarKit.framework/Support/carkitd | PID 34124",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/ManagedConfiguration.framework/Support/profiled | PID 34125",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/UserActivity.framework/Agents/useractivityd | PID 34126",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/Applications/MobileSafari.app/webbookmarksd | PID 34127",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/Frameworks/Accounts.framework/accountsd | PID 34128",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/AssetCacheServices.framework/XPCServices/AssetCacheLocatorService.xpc/AssetCacheLocatorService -d | PID 34129",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/iTunesStore.framework/Support/itunesstored | PID 34130",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/Applications/ServerDocuments.app/PlugIns/ServerFileProvider.appex/ServerFileProvider | PID 34131",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/nsurlstoraged | PID 34132",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/nanoregistryd | PID 34136",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/lsd runAsRoot | PID 34137",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/CalendarDaemon.framework/Support/calaccessd | PID 34138",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/IMDPersistence.framework/XPCServices/IMDPersistenceAgent.xpc/IMDPersistenceAgent | PID 34139",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/sbin/filecoordinationd | PID 34141",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/StreamingZip.framework/XPCServices/com.apple.StreamingUnzipService.xpc/com.apple.StreamingUnzipService | PID 34145",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/nanoregistrylaunchd | PID 34146",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/usr/libexec/sharingd | PID 34150",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/GeoServices.framework/geod | PID 34151",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/Search.framework/searchd | PID 34152",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/Applications/MobileCal.app/PlugIns/MobileCalSpotlight.appex/MobileCalSpotlight | PID 34153",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/Applications/Reminders.app/PlugIns/RemindersSpotlight.appex/RemindersSpotlight | PID 34154",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/Applications/Contacts.app/PlugIns/ContactsCoreSpotlightExtension.appex/ContactsCoreSpotlightExtension | PID 34155",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/FoundationODR.framework/ondemandd.bundle/ondemandd | PID 34159",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/AuthKit.framework/akd | PID 34160",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/Accessibility.framework/Frameworks/AccessibilityUI.framework/XPCServices/com.apple.accessibility.AccessibilityUIServer.xpc/com.apple.accessibility.AccessibilityUIServer | PID 34161",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/Frameworks/PassKit.framework/passd | PID 34162",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/MusicLibrary.framework/Support/medialibraryd | PID 34163",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/Frameworks/AudioToolbox.framework/XPCServices/com.apple.audio.SystemSoundServer-iOS-Simulator.xpc/com.apple.audio.SystemSoundServer-iOS-Simulator | PID 34164",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/PrivateFrameworks/AppSupport.framework/Support/cplogd | PID 34165",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/Applications/MobileCal.app/MobileCal | PID 34166",
    "Process /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk/System/Library/CoreServices/CFNetworkAgent | PID 34168"
), NSUnderlyingError=0x101130010 "The operation couldn’t be completed. (FBSOpenApplicationErrorDomain error 4.)"}
```

**Comment by lawrencelomax** [comment]
1) Are you certain that the arch of the binary matches the Simulator? The `FBSimulatorApplication` instance you create will have an arch property. This should be `i386` or `x86_64` (or both) depending on the target simulator. `FBSimulatorControl` does not currently check archs when launching.
2) Check the Log for the Simulator, there might be something of interest there. Currently this will be stored at `~/Library/Logs/CoreSimulator/Devices/<DEVICE-UDID./system.log`. `FBSimulatorControl` should probably expose this on `FBSimulator`.

**Comment by jaredly** [comment]
Bingo! The simulator arch is x86_64, and my binary is i386. Is there a way to get the simulator to start up as i386? (testing iPad 2, ios 9.0) I didn't find an option for that. I also haven't been able to get xcode to create an x86_64 binary (pretty new to these things, sorry for the noob questions)

**Comment by jaredly** [comment]
Digging more through the code, it looks like there's a separate step for installing the app, which must be done before launching it. Is that the case?

Anyway, I tried
```objc
      BOOL success = [[[session.interact
                        bootSimulator]
                       installApplication:myApp]
                      performInteractionWithError:&error];
```

And this also fails. The simulator boots up fine, but then the "installApplication" step fails, with the error: "Failed to install Application". Digging in, there's an undescriptive error that doesn't end up getting surface that just says "POSIXErrorDomain" with some large integer values.

The system.log shows no log messages when I run my thing. It does show log messages when running via xcode.

Incidentally, if I switch the cli xcode version to 6.4, then I get a different error: "Timeout waiting for device to boot", even though the "iOS Simulator" app opens and boots quickly, long before the timeout error is thrown.

**Comment by lawrencelomax** [comment]
1) Yes you need to install an Application before launching it. I'll update the documentation to clarify this.
2) "Booting" is based on the information that `CoreSimulator` provides to us, not the launch of the Simulator process itself. I don't have 6.4 to hand, but this is the version that I think Travis-CI is currently using.
3) Have you tried on different Simulators? IIRC iPhone 5 may not run with 64 binaries and the 5S was the first iPhone to introduce this. Try running against a later device to confirm that it's definately not an architecture thing.
4) Errors are returned with high-level explainations and the error that is actually manifest from Core Simulator will provide whatever information is available in `NSUnderlyingErrorKey`. The Error Code may provide some value, but it's worth exhausting the possibility that this is an architecture issue first.

**Comment by lawrencelomax** [comment]
Check out [MacErrors.h](http://www.opensource.apple.com/source/CarbonHeaders/CarbonHeaders-18.1/MacErrors.h) for descriptions of the error numbers.

**Comment by oliverhu** [comment]
also happens for me, output is the same as @jaredly 's

**Comment by oliverhu** [comment]
I am able to install the App, but failed to launch it:

2015-10-06 11:49:16.055 xctest[20772:290316] *** Assertion failure in -[FBSimulatorSessionLifecycle simulator:didStartWithProcessIdentifier:terminationHandle:], /Users/khu/ios/FBSimulatorControl/FBSimulatorControl/Session/FBSimulatorSessionLifecycle.m:84
/Users/khu/ios/FBSimulatorControl/FBSimulatorControlTests/Tests/FBSimulatorControlApplicationLaunchTests.m:56: error: -[FBSimulatorControlApplicationLaunchTests testLaunchesSafariApplication] : failed: caught "NSInternalInconsistencyException", "Invalid parameter not satisfying: self.simulatorTerminationHandle == nil"
(
	0   CoreFoundation                      0x00007fff9427b03c __exceptionPreprocess + 172
	1   libobjc.A.dylib                     0x00007fff9015a76e objc_exception_throw + 43
	2   CoreFoundation                      0x00007fff9427ae1a +[NSException raise:format:arguments:] + 106
	3   Foundation                          0x00007fff886d799b -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 195
	4   FBSimulatorControl                  0x000000010526ebd7 -[FBSimulatorSessionLifecycle simulator:didStartWithProcessIdentifier:terminationHandle:] + 743
	5   FBSimulatorControl                  0x00000001052780a6 __46-[FBSimulatorSessionInteraction bootSimulator]_block_invoke + 1462
	6   FBSimulatorControl                  0x0000000105275170 -[FBInteraction_Block performInteractionWithError:] + 80
	7   FBSimulatorControl                  0x00000001052743c1 __35+[FBInteraction chainInteractions:]_block_invoke + 369
	8   FBSimulatorControl                  0x0000000105275170 -[FBInteraction_Block performInteractionWithError:] + 80
	9   FBSimulatorControl                  0x0000000105274fb6 -[FBInteraction performInteractionWithError:] + 70
	10  FBSimulatorControlTests             0x0000000103908294 -[FBSimulatorControlApplicationLaunchTests testLaunchesSafariApplication] + 1236
	11  CoreFoundation                      0x00007fff941527bc __invoking___ + 140
	12  CoreFoundation                      0x00007fff94152612 -[NSInvocation invoke] + 290
	13  XCTest                              0x000000010002134f __24-[XCTestCase invokeTest]_block_invoke_2 + 159
	14  XCTest                              0x0000000100053822 -[XCTestContext performInScope:] + 184
	15  XCTest                              0x000000010002129f -[XCTestCase invokeTest] + 169
	16  XCTest                              0x000000010002173a -[XCTestCase performTest:] + 443
	17  XCTest                              0x000000010001f40b -[XCTestSuite performTest:] + 377
	18  XCTest                              0x000000010001f40b -[XCTestSuite performTest:] + 377
	19  XCTest                              0x000000010000ddbf __25-[XCTestDriver _runSuite]_block_invoke + 51
	20  XCTest                              0x0000000100031f32 -[XCTestObservationCenter _observeTestExecutionForBlock:] + 611
	21  XCTest                              0x000000010000dd08 -[XCTestDriver _runSuite] + 408
	22  XCTest                              0x000000010000e8b7 -[XCTestDriver _checkForTestManager] + 696
	23  XCTest                              0x0000000100054a6f _XCTestMain + 628
	24  xctest                              0x0000000100001dc9 xctest + 7625
	25  libdyld.dylib                       0x00007fff897d05c9 start + 1
	26  ???                                 0x0000000000000005 0x0 + 5
)

**Comment by jaredly** [comment]
I'm now using `xcrun simctl` for installing + running the app, and it works fine. no arch issues

**Comment by lawrencelomax** [comment]
There's probably more that can be done in terms of looking at preconditions before calling certain CoreSimulator methods. I'll look at installed apps shortly. 

@jaredly did you find out what the Error Code corresponded to?

**Comment by oliverhu** [comment]
I manually opened the app and it functions properly, here is the command:

```
    FBSimulatorSession *session = [self.control createSessionForSimulatorConfiguration:[FBSimulatorConfiguration.iPhone6 iOS_8_4] error:&error];
    FBSimulatorApplication *application = [FBSimulatorApplication applicationWithPath:@"/Users/ohu/Library/Developer/Xcode/DerivedData/voyager-frhgmrtfxqiflycndwcgfpqkbhdy/Build/Products/Debug-iphonesimulator/Foo.app" error:&error];
    FBApplicationLaunchConfiguration *appLaunch3 = [FBApplicationLaunchConfiguration
                                                   configurationWithApplication:application
                                                   arguments:@[]
                                                   environment:@{}];
    BOOL install = [[[session.interact bootSimulator] installApplication:application] performInteractionWithError:&error];

    BOOL success = [[[session.interact
                      bootSimulator]
                     launchApplication:appLaunch3]
                    performInteractionWithError:&error];
```

**Comment by oliverhu** [comment]
another question is how to run xctest with this tool? Any sample code for that?

**Comment by lawrencelomax** [comment]
@oliverhu Are you talking about using `FBSimulatorControl` to launch Application Tests on a Simulator? It should be possible to do this for application tests with an Application launch with the `-XCTest` argument.

**Comment by lawrencelomax** [comment]
@oliverhu you should also hopefully see that cryptic message go away after #48 is landed.

**Comment by oliverhu** [comment]
Finally be able to run the application - the issue is I cannot boot simulator again when trying to launch the app.

@lawrencelomax Still not able to run xctest (with host app)

```
    FBSimulatorApplication *application = [FBSimulatorApplication applicationWithPath:@"/Users/ohu/Library/Developer/Xcode/DerivedData/foo-aodwsnztqjhrikgifstschfgfhzf/Build/Products/Debug-iphonesimulator/Foo.app" error:&error];
    FBApplicationLaunchConfiguration *appLaunch3 = [FBApplicationLaunchConfiguration
                                                   configurationWithApplication:application
// also tried @[@"-XCTest"] and @[@"-XCTest", @"All"]
                                                   arguments:@[@"-XCTest", @"All", @"/Users/ohu/Library/Developer/Xcode/DerivedData/foo-aodwsnztqjhrikgifstschfgfhzf/Build/Products/Debug-iphonesimulator/FooTests.xctest"]
                                                   environment:@{}];
    [[[session.interact bootSimulator] installApplication:application] performInteractionWithError:&error];
    [[session.interact launchApplication:appLaunch3] performInteractionWithError:&error];
```

**Comment by foulkesjohn** [comment]
@oliverhu did you manage to get the tests to run? I have managed to get the bundle to load but the tests aren't running...

**Comment by oliverhu** [comment]
@foulkesjohn no ..

**Comment by foulkesjohn** [comment]
I tried moving away from FBSimulatorControl to see if I can get my XCTest bundle to run just using a single simulator. I am getting the same issue, I've put more detail in this SO post: http://stackoverflow.com/questions/33633882/run-xctest-bundle-with-simctl. This probably isn't the place to have this conversation but thought it could help.

**Comment by lawrencelomax** [comment]
I have a [branch that aims to do this](https://github.com/facebook/FBSimulatorControl/commit/5c34403fdf63b85080ba135d4e87ac97657f5322) but haven't had the chance to finish it.

Running `xctest` bundles doesn't require the `xctest` commandline. `xctest` works by injecting a `dylib` into the App-Under-Test that will load the `xctest` bundle at runtime. If you inspect the [system log](https://gist.github.com/lawrencelomax/532d23905a27f97754b7) you can see the environment variables that are passed to the App-Under-Test on launch. In fact this is the  [way that `xctool` launches tests](https://github.com/facebook/xctool/blob/0bc5dfee772075eab02ba26e871a0c76e13b13e5/xctool/xctool/SimulatorWrapper/SimulatorInfo.m#L336) 

cc @ExtremeMan

**Comment by gblotter** [comment]
@lawrencelomax Thanks for the response.  

I've been looking at this thread and #10 's thread.  I checked out the demo project linked in that thread, re-wrote it in Swift and adapted it a bit for my needs, but as I've looked at the system log, I haven't found the right combination of environment variables to use/customize for it to launch an archived xctest bundle.  I'll go check out your branch and see what you're doing.  Thanks for all or your work on this, it's really great stuff.

**Comment by gblotter** [comment]
@lawrencelomax After much digging and inspecting of the system.log, I've not been able to find a way to get it to successfully inject the `dylib` into the App-under-test to have it load the `xctest` bundle at runtime.  I created a demo project doing what I think you're talking about [here](https://github.com/gblotter/FBSimulatorControlDemo).  You can see that I set the arguments and environment params when creating the FBApplicationLaunchConfiguration object.  

I'd appreciate it if you could look at it and tell me what I'm doing wrong.  It sounds like this may have stopped working sometime after the Xcode 7.1 release?

**Comment by foulkesjohn** [comment]
I've looked at xctool and what parameters its passing and brought over the `XCTestConfiguration` stuff. Its hardcoded right now for the bundle but I didn't really have any joy getting tests to run with my test app. Still looks like something is missing but I can't see what: https://github.com/foulkesjohn/FBSimulatorControl/tree/xctest-xctestconfig

**Comment by lawrencelomax** [comment]
It's possible that we could find what is missing in terms of the bootstrapping of the application process by comparing the environment & arguments of an xctest-launched process in #72

I believe that xctool also uses a shim, but I'd need to look to confirm what reason it has this shim for.

**Comment by oliverhu** [comment]
shim should be only a hack to reformat xcode's output to a xctool json format?

**Comment by oliverhu** [comment]
I was able to make xctest (application test) work with FBSimulatorControl. It is a combination of bundle injection settings plus XCTestConfiguration setup. I copied some code from XCTool to make it work, here is the draft commit. 
https://github.com/oliverhu/FBSimulatorControl/commit/929813a10519ceaa6aba8f7e9f85f59e42bdaf72
A demo .. https://youtu.be/uUDDSrC89gM
Also, you would have to wait for the simulator to boot before run the tests.
